WooCommerce/WooCommerceTests/Reviews/ReviewsViewModelTests.swift

```diff
@@ -94,36 +121,85 @@ final class ReviewsViewModelTests: XCTestCase {
 
     func test_hasErrorLoadingData_true_after_failed_sync() {
         // Given
-        viewModel.hasErrorLoadingData = false
+        let mockDataSource = MockReviewsDataSource()
         let storesManager = MockStoresManager(sessionManager: .testingInstance)
         storesManager.whenReceivingAction(ofType: ProductReviewAction.self) { action in
             switch action {
             case .synchronizeProductReviews(_, _, _, _, _, let onCompletion):
-                onCompletion(SampleError.first)
+                onCompletion(.failure(SampleError.first))
             default:
                 return
             }
         }
-        ServiceLocator.setStores(storesManager)
+        let viewModel = ReviewsViewModel(siteID: sampleSiteID, data: mockDataSource, stores: storesManager)
+        viewModel.hasErrorLoadingData = false
 
         // When
         viewModel.synchronizeReviews(pageNumber: 1, pageSize: 25, onCompletion: nil)
 
         // Then
         XCTAssertTrue(viewModel.hasErrorLoadingData)
     }
+
+    func test_synchronizeReviews_triggers_retrieveProducts_with_all_reviewsProductIDs() {
+        // Given
+        let mockDataSource = MockReviewsDataSource()
+        let mocks = MockReviews()
+        let sampleReviews: [ProductReview] = [
+            mocks.review(injectedReviewID: sampleSiteID, injectedProductID: 55),
+            mocks.review(injectedReviewID: sampleSiteID, injectedProductID: 668),
+            mocks.review(injectedReviewID: sampleSiteID, injectedProductID: 789)
```

Comment:
nit: how about adding `GeneratedCopiable` to `ProductReview` and using `copy` with Fakeable here? then we can move away from `MockReviews` (which we can replace with Fakeable and Copeable gradually)

```suggestion
            ProductReview.fake().copy(reviewID: sampleSiteID, productID: 55),
            ProductReview.fake().copy(reviewID: sampleSiteID, productID: 668),
            ProductReview.fake().copy(reviewID: sampleSiteID, productID: 789)
```

Reply:
Good point! Updated in 6de17d180384cafbd14a7d4c0e9a331d5a558ec5